source/extensions/filters/http/grpc_web/grpc_web_filter.cc

```diff
@@ -156,14 +174,43 @@ Http::FilterHeadersStatus GrpcWebFilter::encodeHeaders(Http::ResponseHeaderMap&
   } else {
     headers.setReferenceContentType(Http::Headers::get().ContentTypeValues.GrpcWebProto);
   }
-  return Http::FilterHeadersStatus::Continue;
+
+  if (end_stream || isValidResponseHeaders(headers, end_stream)) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+
+  ENVOY_LOG(debug, "received invalid response headers since the upstream response or local reply "
+                   "is not a gRPC or gRPC-Web response");
+  response_headers_ = &headers;
+  return Http::FilterHeadersStatus::StopIteration;
 }
 
-Http::FilterDataStatus GrpcWebFilter::encodeData(Buffer::Instance& data, bool) {
+Http::FilterDataStatus GrpcWebFilter::encodeData(Buffer::Instance& data, bool end_stream) {
   if (!is_grpc_web_request_) {
     return Http::FilterDataStatus::Continue;
   }
 
+  // When the upstream response (this is also relevant for local reply, since gRPC-Web request is
+  // not a gRPC request which makes the local reply's is_grpc_request set to false) is not a gRPC
+  // response, we set the "grpc-message" header with the upstream body content.
+  if (response_headers_ != nullptr) {
+    if (!end_stream) {
+      return Http::FilterDataStatus::StopIterationNoBuffer;
```

Comment:
won't this cause the same problem?  Shouldn't it buffer until end_stream?

Reply:
Will add a test for this. Seems like I failed to cover this case in this patch.

Reply:
This is now tested.

Reply:
This code should tell the HCM to throw away body data. 
It looks like your test verifies that if you send
a
{\"percentage\": \"100%\"}
that the resultant body is
{\"percentage\": \"100%\"}

I don't think that is the intended behavior?

Reply:
My proposal was: I thought that since this is a bad upstream response, we _can_ throw away part of the body except the last one. Since this is a "bad upstream response" anyway.